### PR TITLE
ENH: approx_fprime --> approx_derivative

### LINKED ISSUE #6026: Replace approx_grad with _numdiff.approx_derivative in scipy.optimize
It would be better to use the more carefully written `approx_derivative` instead of <s>`approx_grad`</s>`approx_fprime` in the various points in scipy.optimize where numerical differentiation is required.

One particular issue is that `approx_grad` does not respect bounds, may evaluate the function outside them, whereas `approx_derivative` has special handling at the boundaries.

### LINKED ISSUE #9042: optimize: Incorrect statement about `jac` in the `minimize` docstring.
In the description of the `jac` argument, the docstring for `scipy.optimize.minimize` says

    Alternatively, the keywords {'2-point', '3-point', 'cs'} select a finite
    difference scheme for numerical estimation of the gradient.

and

    If False, the gradient will be estimated using '2-point' finite difference estimation.

Those descriptions are not correct when `method` is `'Newton-CG'` or `trust-ncg` (and maybe others--I haven't checked).  For example,

```
In [156]: def func(x):
     ...:     return ((x-1)**2).sum()
     ...: 

In [157]: from scipy.optimize import minimize
```

Try `'Newton-CG'` with `jac=False`:

```
In [158]: minimize(func, [1.1, 1.2], method='Newton-CG', jac=False)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-158-94c7580e9508> in <module>()
----> 1 minimize(func, [1.1, 1.2], method='Newton-CG', jac=False)

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    600     elif meth == 'newton-cg':
    601         return _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback,
--> 602                                   **options)
    603     elif meth == 'l-bfgs-b':
    604         return _minimize_lbfgsb(fun, x0, args, jac, bounds,

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/optimize.py in _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback, xtol, eps, maxiter, disp, return_all, **unknown_options)
   1494     _check_unknown_options(unknown_options)
   1495     if jac is None:
-> 1496         raise ValueError('Jacobian is required for Newton-CG method')
   1497     f = fun
   1498     fprime = jac

ValueError: Jacobian is required for Newton-CG method
```

Try `'trust-ncg'` with `jac=False`:

```
In [159]: minimize(func, [1.1, 1.2], method='trust-ncg', jac=False)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-159-0e8fe542e188> in <module>()
----> 1 minimize(func, [1.1, 1.2], method='trust-ncg', jac=False)

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    621     elif meth == 'trust-ncg':
    622         return _minimize_trust_ncg(fun, x0, args, jac, hess, hessp,
--> 623                                    callback=callback, **options)
    624     elif meth == 'trust-krylov':
    625         return _minimize_trust_krylov(fun, x0, args, jac, hess, hessp,

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_trustregion_ncg.py in _minimize_trust_ncg(fun, x0, args, jac, hess, hessp, **trust_region_options)
     32     """
     33     if jac is None:
---> 34         raise ValueError('Jacobian is required for Newton-CG trust-region '
     35                          'minimization')
     36     if hess is None and hessp is None:

ValueError: Jacobian is required for Newton-CG trust-region minimization
```

Apparently a Jacobian *function* is required for these methods.  The `minimize` docstring should be updated to make this clear.  Or, if it is feasible, the methods should be fixed to behave as documented.

If users happen to follow the references to the additional documentation about the [`'Newton-CG` method](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-newtoncg.html#optimize-minimize-newtoncg), they'll see the note that says "Note that the jac parameter (Jacobian) is required."  That's too well hidden, and still not explicit.  Does it mean providing the argument `jac='3-point'` is sufficient?  If an implementation of the Jacobian function (i.e. a callable) is required, it should say so explicitly, and that information should be in the `minimize` docstring.

See https://stackoverflow.com/questions/33926357/jacobian-is-required-for-newton-cg-method-when-doing-a-approximation-to-a-jaco and https://stackoverflow.com/questions/51369984/jacobian-is-required-for-newton-cg-method-issue-still-exists-for-trust-ncg for cases where the current documentation leads to confusion.

### PR DESCRIPTION
**Pull request by andyfaff**
This PR transfers the calculation of function derivatives from `_approx_fprime_helper` to `approx_derivative`.
Changes made:

- `approx_fprime` is now a wrapper around `approx_derivative`.
- LBFGSB gradient calculation is done directly `approx_derivative`, taking into account the LBFGSB bounds. This should close a few issues where the gradient calculation inadvertently steps outside the bounds.
- <s>a `_compute_relative_step` function is added to calculate relative steps from absolute steps, because `approx_fprime` uses absolute steps.</s>

Future changes suggested:

- BFGS and CG should use ScalarFunction, this would save a few function evaluations, and be cleaner. However, they currently use absolute finite difference steps, whilst ScalarFunction uses relative steps. I don't know of any suggestions where it would be better to use absolute differences, because relative steps should allow the step to be scaled correctly.
- allow all finite difference minimizers to choose between 2 and 3-point finite differences

closes #9042
closes #6026
closes #9634

### CODE REVIEW

**Comment by andyfaff** [comment]
The PR is easiest to review by looking at the commits separately.

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -761,7 +737,20 @@ def approx_fprime(xk, f, epsilon, *args):
     array([   2.        ,  400.00004198])
 
     """
-    return _approx_fprime_helper(xk, f, epsilon, args=args)
+    xk = np.asarray(xk, float)
+
+    rel_step = _compute_relative_step(epsilon, xk, '2-point')
+    f0 = f(*((xk,) + args))
```
Comment (peterbell10) [inline scipy/optimize/optimize.py]: Now that python 3.5 is the minimum version, I think you can use generalised argument unpacking
```suggestion
    f0 = f(xk, *args)
```

**Comment by andyfaff** [comment]
I decided to press on with updating various methods to use `approx_derivative`. This PR will need #10693 merged first, otherwise a lot of tests will fail.

**Comment by andyfaff** [comment]
Also does half of #5952

**Comment by andyfaff** [comment]
(opened/closed because recent commit wasn't showing up)

**Comment by antonior92** [comment]
It is really great that you have fixed that!

It didn't made much sense to have several different implementations of finite differences, with some methods still using the old one...

File: scipy/optimize/_minimize.py [hunk 2]
```diff
@@ -531,29 +535,25 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
...
+        pass
+    elif jac is True:
+        # fun returns func and grad
+        fun = MemoizeJac(fun)
+        jac = fun.derivative
+    elif (jac in ('2-point', '3-point', 'cs') and
+          meth in ['trust-constr', 'bfgs', 'cg', 'l-bfgs-b', 'tnc']):
+        # finite differences
+        pass
+    elif meth in ['trust-constr']:
```
Comment (antonior92) [inline scipy/optimize/_minimize.py]: It has been a long time since I last looked into that. But I understood that '2-point' and None is the same in the end. 

Could we colapse the last 2 elif and the last else into just:
```
else:
    jac = None
```
??

Comment (antonior92) [reply scipy/optimize/_minimize.py]: Because it seems to me that the result would be the same as:
```
elif meth in ['trust-constr']:
       # default jac calculation for this method
        jac = '2-point'
elif jac is None or bool(jac) is False:
    # this will cause e.g. LBFGS to use forward difference, absolute step
    jac = None
else
    # default if jac option is not understood
     jac = None
```

File: scipy/optimize/_minimize.py [hunk 3]
```diff
@@ -531,29 +535,25 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
...
-            jac = '2-point'
-        elif not callable(jac) and jac not in ('2-point', '3-point', 'cs'):
-            raise ValueError("Unsupported jac definition.")
+    if callable(jac):
+        pass
+    elif jac is True:
+        # fun returns func and grad
+        fun = MemoizeJac(fun)
+        jac = fun.derivative
+    elif (jac in ('2-point', '3-point', 'cs') and
```
Comment (ev-br) [inline scipy/optimize/_minimize.py]: Maybe add a constant tuple of allowed FD methods to _numdiff.py? This way, if a new scheme is added to approx_derivative at some point, we won't need to update it here

File: scipy/optimize/_minimize.py (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/_minimize.py]: >  But I understood that '2-point' and None is the same in the end.

You're correct for the existing behaviour. Before this PR if the method was `'2-point'` and the method wasn't 'trust-constr' `jac` would be [converted](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L550) to `None`. 
In all the `_minimize_*` methods that would correspond to using forward differences with an *absolute* step using `approx_fprime`. 
If the 'trust_constr' method was used `jac in {None, '2-point'}` would be done via forward differences with a *relative* step in `approx_derivative`.

In this PR we want to ensure back compatibility. So if:
- `jac=None` the gradient will still be calculated using forward differences with an *absolute* step in `approx_derivative`.
- `jac=None` and `meth='trust-constr`, still needs forward differences with a *relative* step, so make `jac='2-point'`. If `None` is sent to 'trust-constr' then the method won't understand it (see [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py#L331), then [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentiable_functions.py#L31))
- The `else` is equivalent to arriving at this [line](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L556). I think I tried removing the else and there were test failures.

The new behaviour is that minimizer methods can now use `'2-point', '3-point', 'cs'`, via `approx_derivative` and a *relative step*. In other words:
- `jac=None`-->absolute step with forward differences
- `jac in {'2-point', '3-point', 'cs'}` --> relative step. 
Relative step is probably better in nearly all cases. If the latter is employed then a newly added `finite_diff_rel_step` keyword is used to figure out what the relative step is. This keyword naming followed 'trust-constr'.

Currently there is no way of the minimizer methods asking for `2-point`, `3-point`, or `cs` with an absolute step. This would require an extra keyword of the form `step_type={'relative', 'absolute'}`. I figured this change could be made in the future if absolutely required, but at the expense of adding another keyword.
(note that `approx_derivative` could handle either calculation).

File: scipy/optimize/_minimize.py (hunk 3 above)
Comment (andyfaff) [reply scipy/optimize/_minimize.py]: This already [exists](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentiable_functions.py#L9), can make the change.

Comment (andyfaff) [reply scipy/optimize/_minimize.py]: Done.

File: scipy/optimize/lbfgsb.py [hunk 4]
```diff
@@ -275,27 +286,51 @@ def _minimize_lbfgsb(fun, x0, args=(), jac=None, bounds=None,
...
+            g = approx_derivative(func, x, abs_step=eps, f0=f,
+                                  method='2-point', bounds=new_bounds)
             return f, g
-    else:
+
+    elif callable(jac):
         def func_and_grad(x):
-            f = fun(x, *args)
-            g = jac(x, *args)
+            return func(x), jac(x, *args)
```
Comment (ev-br) [inline scipy/optimize/lbfgsb.py]: `func(x,*args)`?

Comment (andyfaff) [reply scipy/optimize/lbfgsb.py]: `func` is [`fun` wrapped with `args`](https://github.com/scipy/scipy/pull/10673/files#diff-be9847790086f0afe52f687cc3ea9038R305)

File: scipy/optimize/_minimize.py (hunk 2 above)
Comment (antonior92) [reply scipy/optimize/_minimize.py]: Ahh ok. Good that you considered that.

You can probably merge the else with last elif though, since both set `jac = None`

**Comment by andyfaff** [comment]
Rebased and squashed because the commits were getting unwieldy.

**Comment by andyfaff** [comment]
This PR is in a fairly good state at the moment, but there are two doctest failures.
The first is a non-reproducibility:
```
scipy.optimize.dual_annealing
-----------------------------
File "build/testenv/lib/python3.6/site-packages/scipy/optimize/_dual_annealing.py", line 595, in dual_annealing
Failed example:
    print("global minimum: xmin = {0}, f(xmin) = {1:.6f}".format(
          ret.x, ret.fun))
Expected:
    global minimum: xmin = [-4.26437714e-09 -3.91699361e-09 -1.86149218e-09 -3.97165720e-09
     -6.29151648e-09 -6.53145322e-09 -3.93616815e-09 -6.55623025e-09
    -6.05775280e-09 -5.00668935e-09], f(xmin) = 0.000000
Got:
    global minimum: xmin = [ 4.39135632e-09 -7.56854527e-09 -4.87609759e-09  4.10762504e-09
     -4.55194300e-09 -5.28219832e-09 -3.64137371e-09  5.11390148e-09
      3.88893317e-09 -2.68683971e-09], f(xmin) = 0.000000
```
I think this is the doctest that repeatedly gives issues. The same result is found, but it doesn't meet the requirements of whatever does the comparison. I think in this case the local minimizer for `dual_annealing` is going to be LBFGSB because bounds are in use. LBFGSB is touched by this PR, but the test suite passes.

The second failure is:
```
scipy.optimize.shgo
-------------------
File "build/testenv/lib/python3.6/site-packages/scipy/optimize/_shgo.py", line 399, in shgo
Failed example:
    res
Expected:
         fun: 29.894378159142136
        funl: array([29.89437816])
     message: 'Optimization terminated successfully.'
        nfev: 119
         nit: 3
       nlfev: 40
       nlhev: 0
       nljev: 5
     success: True
           x: array([6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02])
          xl: array([[6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02]])
Got:
         fun: 29.894378159142136
        funl: array([29.89437816])
     message: 'Optimization terminated successfully.'
        nfev: 114
         nit: 3
       nlfev: 35
       nlhev: 0
       nljev: 0
     success: True
           x: array([6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02])
          xl: array([[6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02]])
```
By default shgo uses SLSQP as a local minimizer. The differing output is down to SLSQP's behaviour before and after this PR:

before
-------
SLSQP includes a finite difference estimation of gradient in the total number of gradient evaluations (`OptimizeResult.njev > 0`).

after
-----
SLSQP excludes a finite difference estimation of gradient in the total number of gradient evaluations (`OptimizeResult.njev == 0`).
This is because `ScalarFunction` does not count a finite difference evaluation of gradient as increasing the number of gradient calculations. `njev` is increased if a callable jacobian is provided.

Not sure what to do here because the all the minimize methods currently practice a mixture of the two. Advice from other devs is welcomed:

- should `njev` include gradient calculations done with finite differences (maintaining 'back compatibility' as currently exists for SLSQP/BFGS/CG/TNC/LBFGSB/etc)?
- *or* should it only include njev calculations from a callable jacobian (as currently exists for dogleg/trust-constr/trust-ncg/trust-exact/trust-krylov)?

The first option would need to bring 'trust-*' and 'dogleg' into line with the others. The second option would bring SLSQP/BFGS/etc into line with 'trust-*' and 'dogleg'. I prefer the second option, but both are doable.

Otherwise the end results for all the minimizers are the same. The caching behaviour of ScalarFunction has in some circumstances reduced the total number of function evaluations required to reach a solution.

**Comment by mdhaber** [comment]
> Not sure what to do here. Advice from other devs is welcomed:
should njev include gradient calculations done with finite differences?
or should it only include njev calculations from a callable jacobian?

Without reading the docs, I would probably assume that `njev` would report the number of times the gradient was calculated or approximated by finite differences, so the first option.

As for failing doctests, there is precedent for adding `# may vary` on the lines that might be different, e.g. gh-7652. In that case, I would suggest adding the doctest into the test suite (also) and doing the comparison in a more appropriate way. (I think doctests need to match verbatim.)

**Comment by ev-br** [comment]
The shgo doctest failure reports the same `nit` but different `n1jev` and `nfev`. 

The dual_annealing failure failing doctest has prints, and that forces the checker to fall back to plain doctest. Consider changing the print to 
```>>> ret.x
<Value>
>>> ret.fun
<value>
```

**Comment by rgommers** [comment]
This needs a rebase. Would be nice to get this into 1.4.0 and looks close to ready, but it's a big PR. If anyone who already reviewed would want to take another look, that'd be very useful.

**Comment by andyfaff** [comment]
@rgommers rebased. Lets see if it comes back green.

**Comment by andyfaff** [comment]
Whilst this probably won't get in for 1.4, it would be good to get a decision on whether it's a Go/No Go for this PR. @nmayorov @mdhaber @ev-br @antonior92 @rgommers. as far as I can tell the only factor in this is #10808. That issue discusses whether FD should be done outside minimize.
I think the two are separated somewhat. This PR updates what is currently done inside minimize and brings each optimizer to a similar level. Any changes forthcoming from 10808 should be back compatible.

**Comment by rgommers** [comment]
> should `njev` include gradient calculations done with finite differences?

looks like you went with this, which seems like the right choice to me

> That issue discusses whether FD should be done outside minimize.
> I think the two are separated somewhat. This PR updates what is currently done inside minimize and brings each optimizer to a similar level. Any changes forthcoming from 10808 should be back compatible.

This makes sense. All changes here are private, it seems like a nice refactoring that doesn't block any future changes.

I just went through the whole PR and it looks good. Modulo the one open comment on `2-point`/`None` I think it's about ready to merge.

**Comment by andyfaff** [comment]
Having revisited @antonior92, I agree that the last `elif/else` could be condensed. However, I'd prefer to keep it as it is. This is because the code then explicitly: does what the docstring says, outlines the behaviour if `jac is None or bool(jac) is False`. This means the codebase keeps a memory of what the code used to do before this PR, and might make keeping back compatbility easier later on.

**Comment by tylerjereddy** [comment]
I will bump to 1.5.0, but let's try to wrap this up soon since it is so close it seems.

**Comment by ev-br** [comment]
+1 from me. Maybe best to merge early in the release cycle.

File: scipy/optimize/optimize.py [hunk 5]
```diff
@@ -762,7 +830,19 @@ def approx_fprime(xk, f, epsilon, *args):
     array([   2.        ,  400.00004198])
 
     """
-    return _approx_fprime_helper(xk, f, epsilon, args=args)
+    xk = np.asarray(xk, float)
+
+    f0 = f(xk, *args)
+    if not np.isscalar(f0):
```
Comment (eric-wieser) [inline scipy/optimize/optimize.py]: This condition, and the .item() below, looks wrong to me - I think you want:
```
if np.ndim(f0) != 0:
    raise ValueError(...)
```

File: scipy/optimize/optimize.py [hunk 6]
```diff
@@ -984,29 +1065,43 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
...
-        grad_calls, myfprime = wrap_function(approx_fprime, (f, epsilon))
-    else:
-        grad_calls, myfprime = wrap_function(fprime, args)
+    f = sf.fun
+    myfprime = sf.grad
+
+    old_fval = f(x0)
     gfk = myfprime(x0)
+
+    if not np.isscalar(old_fval):
```
Comment (eric-wieser) [inline scipy/optimize/optimize.py]: Same as above - isscalar is usually the wrong approach.

Comment (andyfaff) [reply scipy/optimize/optimize.py]: For better or for worse I think the historical behaviour of most minimizers was to work with an objective function `f` that returned any of `1.1`, `np.array(1.1)`, `np.array([1.1])`. What we need to filter out is `np.array([1.1, 2.2])`, etc. The code as it's currently written does what is required.

One could change to:

```
if np.ndim(df):
    try:
        # required to allow np.array([1.1]) to work.
        df = df.item()
    except ValueError:
            raise ValueError("The user-provided "
                             "objective function must "
                             "return a scalar value.") 
```

However, this has it's own problems; If a function doesn't have an explicit return then it's return value is None:

```
>>> np.ndim(None)
0
>>> np.isscalar(None)
False
```

There are several test cases that don't work if the `np.array([1.1])` is returned by the objective and if `df.item()` is removed, 7dd14d567a8248be3956fd09cb822b0dde502a1. `df.item()` converts df to a standard Python scalar. If `np.array([1.1])` is not permitted at this point I think that would cause quite a few issues downstream as back compatibility would be broken.

What are the use cases where the current behaviour causes problems for users?

Comment (eric-wieser) [reply scipy/optimize/optimize.py]: Your commit ref there doesn't seem to have linked, sadly.

Using isscalar will rule out other number types like Decimal and Fraction, which doesn't look like part of the intent of this patch.

There are very few problems for which `isscalar` is an appropriate solution. Will follow up later on the dangers of item

Comment (andyfaff) [reply scipy/optimize/optimize.py]: Hopefilly this works: https://github.com/scipy/scipy/commit/07dd14d567a8248be3956fd09cb822b0dde502a1#diff-c1c2c73ad9e5ddef5673e2e95d19020f

Comment (andyfaff) [reply scipy/optimize/optimize.py]: Would coercion to a float, `df = float(df)`, be acceptable?

**Comment by pv** [comment]
float() is probably the right thing if you want a float from user-provided scalar objective function return value.

File: scipy/optimize/optimize.py (hunk 6 above)
Comment (eric-wieser) [reply scipy/optimize/optimize.py]: Ideally I'd go for:
```
f0 = asarray(f0)
if f0.ndim != 0:
    try:
        f0 = f0.squeeze(axis=range(f0.ndim))
    except ValueError:
        raise ValueError("The user-provided ...")
    warnings.warn(UserWarning, "User-provided function returned an array, which was cast to a scalar", stacklevel=2)
```

Comment (eric-wieser) [reply scipy/optimize/optimize.py]: Ah, I now see this is just moved code.

Why does the check not live in `approx_derivative`?

Comment (andyfaff) [reply scipy/optimize/optimize.py]: That code is fairly understandable, but I can't see why a simple float cast wouldn't work

File: scipy/optimize/optimize.py [hunk 7]
```diff
@@ -164,6 +166,97 @@ def vecnorm(x, ord=2):
...
+        # absolute step size. ScalarFunction has to be provided an
+        # epsilon value that is not None to use absolute steps. This is
+        # normally the case from most _minimize* methods.
+        grad = '2-point'
+        epsilon = epsilon
+
+    fhess = hess
+    if hess is None:
+        def fhess(x, *args):
+            return x
```
Comment (eric-wieser) [inline scipy/optimize/optimize.py]: nit: shorter as:
```python
    if hess is None:
        def hess(x, *args):
            return x
```

File: scipy/optimize/optimize.py (hunk 6 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: In BFGS the user can provide a jacobian, so a test for scalar-ness is not necessarily done in `approx_derivative`, or in `_differentiable_functions.ScalarFunction`

Comment (eric-wieser) [reply scipy/optimize/optimize.py]: @andyfaff: a float cast would downcast `longdouble`, in case users want more precision

Comment (eric-wieser) [reply scipy/optimize/optimize.py]: > In BFGS the user can provide a jacobian, so approx_derivative is not necessarily called.

Apologies, that comment was intended for the other instance of this in `approx_fprime`

Comment (andyfaff) [reply scipy/optimize/optimize.py]: > a float cast would downcast longdouble and Decimal

The stumbling block there is the [numerical differentiation code is designed to work in float](https://github.com/scipy/scipy/blob/master/scipy/optimize/_numdiff.py#L12).

Comment (andyfaff) [reply scipy/optimize/optimize.py]: `approx_fprime` isn't really used for anything after this PR. It's just the public API for a simple scalar function differentiation.
The check is retained here because `approx_derivative` can differentiate functions returning non-scalar arrays, but we want to limit `approx_fprime` to the scalar case only.

Comment (eric-wieser) [reply scipy/optimize/optimize.py]: If nothing else, it seems sloppy to convert back and forth between numpy scalars and python scalars

**Comment by andyfaff** [comment]
No changes made in the last commit, just resolving merge conflicts

File: scipy/optimize/optimize.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: ACK

**Comment by andyfaff** [comment]
pinging for a possible merge if people are happy.

**Comment by ev-br** [comment]
Let's get this in now, and iron out possible wrinkles downstream while there's time before the next release. Thanks Andrew, all!

**Comment by tylerjereddy** [comment]
This PR automatically closed another PR (#9634)---just double checking that was intended?

**Comment by andyfaff** [comment]
@tylerjereddy in https://github.com/scipy/scipy/pull/10673/files#diff-be9847790086f0afe52f687cc3ea9038R393

**Comment by mdhaber** [comment]
Thanks @andyfaff. Could you add this contribution to the 1.5.0 [release notes](https://github.com/scipy/scipy/wiki/*Release-note-entries-for-SciPy-1.5.0*)?

**Comment by mdhaber** [comment]
@andyfaff it looks like you added these? Just wanted to be sure this is "The finite difference numerical differentiation used in various `minimize` methods that use gradients has several new features:"

**Comment by andyfaff** [comment]
Yes.

**Comment by austinwiltshire** [comment]
I may be seeing an issue introduced in these changes, related to work in the skopt package (https://github.com/scikit-optimize/scikit-optimize) if you want to run their unit tests directly. I'll try and recreate a test case here.

I'm not numerically minded, so I have no idea if perhaps skopt is wrong since it is using scipy's optimize.approx_f_prime as a reference to compare it's own optimization. However, this test worked up until scipy 1.5, meaning at least the outputs of this function changed (maybe y'all fixed a bug. Just reporting in case one was introduced instead).

I'm trying to capture a lot of this test that's now failing to see if its helpful. I am including the reference to some code in skopt as I figured trying to copy paste the actual function being tested would butcher it. You can see Matern's __call__ function here implementation here: https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.Matern.html
https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/gaussian_process/kernels.py#L1464


```
from sklearn.gaussian_process.kernels import Matern

kernel = Matern(length_scale=[1, 2, 3, 4, 5], nu=0.5)

def kernel_X_Y(x, y, kernel):
    X = np.expand_dims(x, axis=0)
    Y = np.expand_dims(y, axis=0)
    return kernel(X, Y)[0][0]

X = [1.6243453636632417, -0.6117564136500754, -0.5281717522634557, -1.0729686221561705, 0.8654076293246785]
step_size = 1e-06
y = [1.6243453636632417, -0.6117564136500754, -0.5281717522634557, -1.0729686221561705, 0.8654076293246785]

expected_answer = np.ndarray([-0.9999995, -0.49999988, -0.33333328, -0.24999997, -0.19999998])
answer = optimize.approx_fprime(X, kernel_X_Y, step_size, y, kernel)
```

In the above, answer currently equals np.ndarray([-0.9999995,   0.49999988,  0.33333328,  0.24999997, -0.19999998])

The signs on the second, third and fourth argument are all flipped from what they were in the 1.4.1 version of this function, in other words.

While the issue seems to be with sklearn's Matern kernel and scipy's optimize.approx_fprime, the actual test is drawn from here:

https://github.com/scikit-optimize/scikit-optimize/blob/907ec9d0b82a6abaeed474944d76d98f3407af18/skopt/learning/gaussian_process/tests/test_kernels.py#L90

I'm using numpy 1.19, scikit-learn 0.23.1 and master of skopt. The test passes on scipy 1.14.1 and fails at 1.15

**Comment by WarrenWeckesser** [comment]
A quick plot of the cross-section through that Matérn kernel shows that the first derivative is not continuous:
```
def kernel_cross_section(t, k, x, y, kernel):
    x = x.copy()
    result = []
    for tval in t:
        x[k] = tval
        result.append(kernel_X_Y(x, y, kernel))
    return np.array(result)

k = 1
t = np.linspace(0.9*X[k], 1.1*X[k], 1000)
f = kernel_cross_section(t, k, X, y, kernel)

plt.plot(t, f)
plt.axvline(X[k], linestyle='--', alpha=0.3)
plt.title(f'Cross-section index {k}')
plt.show()
```

Figure_1


According to  the [docstring for `approx_fprime`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.approx_fprime.html), 

> the function gradient is determined by the forward finite difference formula:

```
         f(xk[i] + epsilon[i]) - f(xk[i])
f'[i] = ---------------------------------
                    epsilon[i]
```

All the cross-sections have a peak like the one shown above.  If the finite differences used by `approx_fprime` are, in fact, forward finite difference as stated and shown in the formula in the docstring, `approx_fprime` shouldn't sample the function to the left of the discontinuity.  For the Matérn kernel, `approx_fprime` should return only negative values.  So the old values returned by `approx_fprime` appear to be the correct values.

In short, this appears to be a regression in `approx_fprime`.

**Comment by andyfaff** [comment]
@austinwiltshire could you open an issue for this please?  
The internals of approx_fprime now use approx_derivative, it may be that that does backwards differences by default.

**Comment by andyfaff** [comment]
Which would be an issue at a discontinuity.

**Comment by WarrenWeckesser** [comment]
@andyfaff, I created an issue for this: https://github.com/scipy/scipy/issues/12487
